pkg/querier/batch/stream.go

```diff
@@ -39,64 +45,120 @@ func (bs *batchStream) atTime() int64 {
 
 func (bs *batchStream) at() (int64, float64) {
 	b := &(*bs)[0]
-	return b.Timestamps[b.Index], b.Values[b.Index]
+	return b.Timestamps[b.Index], b.SampleValues[b.Index]
+}
+
+func (bs *batchStream) atHistogram() (int64, *histogram.Histogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.HistogramValues[b.Index]
 }
 
+func (bs *batchStream) atFloatHistogram() (int64, *histogram.FloatHistogram) {
+	b := &(*bs)[0]
+	return b.Timestamps[b.Index], b.FloatHistogramValues[b.Index]
+}
+
+// mergeStreams merges streams of Batches of the same series over time.
+// Samples are simply merged by time when they are the same type (float/histogram/...), with the left stream taking precedence if the timestamps are equal.
+// When sample are different type, batches are not merged. In case of equal timestamps, histograms take precedence since they have more information.
 func mergeStreams(left, right batchStream, result batchStream, size int) batchStream {
+
 	// Reset the Index and Length of existing batches.
 	for i := range result {
 		result[i].Index = 0
 		result[i].Length = 0
 	}
+
 	resultLen := 1 // Number of batches in the final result.
 	b := &result[0]
 
-	// This function adds a new batch to the result
-	// if the current batch being appended is full.
-	checkForFullBatch := func() {
+	// Step to the next Batch in the result, create it if it does not exist
+	nextBatch := func(valueTypes chunkenc.ValueType) {
+		// The Index is the place at which new sample
+		// has to be appended, hence it tells the length.
+		b.Length = b.Index
+		resultLen++
+		if resultLen > len(result) {
+			// It is possible that result can grow longer
+			// then the one provided.
+			result = append(result, createBatch(valueTypes))
+		}
+		b = &result[resultLen-1]
```

Comment:
Why don't we directly set the `b` type here to `valueTypes`? The function `nextBatch()` looks inconsistent to me, because the input `valueTypes` is honored only if a new batch is created, but not reused.